// gojs-angular/esm2015/lib/palette.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/palette.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, NgZone, Output, ViewChild } from '@angular/core';
import * as go from 'gojs';
import { NgDiagramHelper } from './ng-diagram-helper';
export class PaletteComponent {
    /**
     * @param {?} zone
     */
    constructor(zone) {
        this.zone = zone;
        /**
         * Link data for palette. Optional.
         */
        this.linkDataArray = null;
        /**
         * Model data for palette. Optional.
         */
        this.modelData = null;
        /**
         * Event emitter -- fires when palette model changes. Capture this emitted event in parent component
         */
        this.modelChange = new EventEmitter();
        /**
         * The Palette itself
         */
        this.palette = null;
    }
    /**
     * Initialize Palette after view init
     * @return {?}
     */
    ngAfterViewInit() {
        if (!this.paletteDiv) {
            throw new Error("paletteDiv is not defined");
        }
        this.palette = this.initPalette();
        if (!(this.palette instanceof go.Palette)) {
            throw new Error("initPalette function did not return a go.Palette");
        }
        // reduces change detection on mouse moves, boosting performance
        NgDiagramHelper.makeMouseMoveRunOutsideAngularZone(this.palette, this.zone);
        // assign the Palette's div, which (among many other things) will attach a bunch of listeners to the canvas,
        // using the overridden addEventListener function above
        /** @type {?} */
        const divRef = this.paletteDiv.nativeElement;
        if (divRef == null)
            return;
        this.palette.div = divRef;
        // initialize palette model
        NgDiagramHelper.initializeModel(this.palette, this.nodeDataArray, this.linkDataArray, this.modelData);
    }
    /**
     * If a change has occured on an \@Input property, merge the app-level changes with GoJS
     * @return {?}
     */
    ngOnChanges() {
        if (!this.palette || !this.palette.model)
            return;
        NgDiagramHelper.mergeAppDataWithModel(this);
    } // end ngOnChanges 
    // end ngOnChanges 
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.palette.div = null; // removes event listeners
    }
}
PaletteComponent.decorators = [
    { type: Component, args: [{
                selector: 'gojs-palette',
                template: '<div #ngPalette [className]=divClassName></div>'
            }] }
];
/** @nocollapse */
PaletteComponent.ctorParameters = () => [
    { type: NgZone }
];
PaletteComponent.propDecorators = {
    initPalette: [{ type: Input }],
    nodeDataArray: [{ type: Input }],
    linkDataArray: [{ type: Input }],
    modelData: [{ type: Input }],
    divClassName: [{ type: Input }],
    modelChange: [{ type: Output }],
    paletteDiv: [{ type: ViewChild, args: ['ngPalette', { static: true },] }]
};
if (false) {
    /**
     * Palette initialization function. Returns a go.Palette.
     * Do not initialize model data in this function.
     * @type {?}
     */
    PaletteComponent.prototype.initPalette;
    /**
     * Node data for palette
     * @type {?}
     */
    PaletteComponent.prototype.nodeDataArray;
    /**
     * Link data for palette. Optional.
     * @type {?}
     */
    PaletteComponent.prototype.linkDataArray;
    /**
     * Model data for palette. Optional.
     * @type {?}
     */
    PaletteComponent.prototype.modelData;
    /**
     * Palette div class name. Use this name to style your palette in CSS
     * @type {?}
     */
    PaletteComponent.prototype.divClassName;
    /**
     * Event emitter -- fires when palette model changes. Capture this emitted event in parent component
     * @type {?}
     */
    PaletteComponent.prototype.modelChange;
    /**
     * The DIV element holding the Palette
     * @type {?}
     */
    PaletteComponent.prototype.paletteDiv;
    /**
     * The Palette itself
     * @type {?}
     */
    PaletteComponent.prototype.palette;
    /** @type {?} */
    PaletteComponent.prototype.zone;
}
